tapgarden/caretaker.go

```diff
@@ -216,15 +220,29 @@ func (b *BatchCaretaker) Cancel() CancelResp {
 				"cancel failed: %w", batchKey, batchState, err)
 		}
 
-		b.cfg.BroadcastErrChan <- fmt.Errorf("caretaker canceled")
-
-		return CancelResp{&finalBatchState, err}
+		cancelResp = CancelResp{&finalBatchState, err}
 
 	default:
 		err := fmt.Errorf("BatchCaretaker(%x), batch not cancellable",
 			b.cfg.Batch.BatchKey.PubKey.SerializeCompressed())
-		return CancelResp{nil, err}
+		cancelResp = CancelResp{nil, err}
 	}
+
+	b.cfg.CancelRespChan <- cancelResp
+
+	// If the batch was cancellable, the finalState of the cancel response
+	// will be non-nil. If the cancellation failed, that error will be
+	// handled by the planter. At this point, the caretaker should always
+	// shut down gracefully.
+	if cancelResp.finalState != nil {
```

Comment:
It seems to me that `finalState` is used indirectly to indicate whether the batch was successfully cancelled or not.

Perhaps we should add a new bool flag to `CancelResp` called `BatchCancelSuccessful` or something like that which would handle that state explicitly. This might help us get rid of some comments.

Reply:
Good catch! Reading this again the planter and the other callsites for `Cancel()` don't actually need the final batch state, just if the batch was cancellable. So `finalState` could be replaced with a bool.

We still want to report any potential errors from the DB update.